borgmatic/hooks/pushover.py

```diff
@@ -0,0 +1,89 @@
+import logging
+
+import requests
+
+logger = logging.getLogger(__name__)
+
+
+def initialize_monitor(
+    ping_url, config, config_filename, monitoring_log_level, dry_run
+):  # pragma: no cover
+    '''
+    No initialization is necessary for this monitor.
+    '''
+    pass
+
+
+def ping_monitor(hook_config, config, config_filename, state, monitoring_log_level, dry_run):
+    '''
+    Post a message to the configured Pushover application.
+    If this is a dry run, then don't actually update anything.
+    '''
+
+    run_states = hook_config.get('states', ['fail'])
+
+    if state.name.lower() not in run_states:
+        return
+
+    dry_run_label = ' (dry run; not actually updating)' if dry_run else ''
+
+    state_config = hook_config.get(
+        state.name.lower(),
+        {
+            'message': state.name.lower(),
+        },
+    )
+
+    token = hook_config.get('token')
+    user = hook_config.get('user')
+
+    logger.info(f'{config_filename}: Updating Pushover {dry_run_label}')
+
+    if token is None:
+        logger.warning(f'{config_filename}: Token missing for Pushover')
+        return
+    if user is None:
+        logger.warning(f'{config_filename}: User missing for Pushover')
+        return
+
+    if 'priority' in state_config and state_config['priority'] == 2:
+        if 'expire' not in state_config:
+            logger.info(f'{config_filename}: Setting expire to default (10min).')
+            state_config['expire'] = 1200
+        if 'retry' not in state_config:
+            logger.info(f'{config_filename}: Setting retry to default (30sec).')
+            state_config['retry'] = 30
+    else:
+        state_config.pop('expire', None)
+        state_config.pop('retry', None)
```

Comment:
So if I'm reading this section correctly, `expire` and `retry` are ignored unless `priority` is 2? And they're set to default values only when `priority` is 2? Rather than silently swallowing these values, it might be good to raise an error because it's basically a misconfiguration.

Another option would be to push this logic into the schema via [`oneOf`](https://json-schema.org/understanding-json-schema/reference/combining#oneOf) (e.g., so that there are two sub-schemas, one with `priority` with only the choice of 2 and another schema with `priority` only 1 or below). But that might be more complicated JSON Schema work than you want to undertake now.

Reply:
Hi, since expire and retry are only used on priority 2 (emergency), I wrote the logic to remove them if they are delared when the priority is not set to 2.

Would you be ok with me changing it to continue doing what it is doing however give a warning that says something like: "Expire and/or Retry should only be configured when setting Priority to 2. They will be removed from the request." ?

Reply:
That's fine with me, although what's the use case for allowing a user to specify `expire`/`retry` when `priority` != 2? Like why not just error?

Reply:
My thought was to make it a bit more forgiving to someone who is setting up the hook. Instead of not working, we'd warn them.

I don't feel strongly either way so I can change it if you'd like 😁

Reply:
I don't feel strongly but I generally err towards erroring on misconfigurations (as opposed to runtime issues), but I could see the merits of both approaches!

Reply:
I will keep with the standard practice then and error out. Logging an error and returning is ok?

Reply:
That actually won't cause execution to stop. So if you want to error instead of warn, I would instead suggest raising `ValueError`, which borgmatic will catch higher up and display nicely. See `normalize.py` if you want an example of raising that exception type for a similar purpose.

Reply:
I am getting close. I have committed what seems to be working code https://github.com/borgmatic-collective/borgmatic/pull/86/commits/a939a66fb44421488332a5060164cefc7335fb50 ... however my test `test_ping_monitor_start_state_backup_default_message_with_priority_high_declared_expire_and_retry_delared_success` now fails. I cannot figure out how to get flexmock to look for a ValueError.

Any suggestions?

Reply:
The magical incantation is:

```python
with pytest.raises(ValueError):
    module.ping_monitor(
        ...
    )
```

See `tests/unit/hooks/test_mariadb.py:test_dump_data_sources_errors_for_missing_all_databases()` for an example.

Reply:
Sorry for the delay. Here is the test. Thanks for the pointer, I was stuck on that since I wasn't sure pytest was allowed to be included and I kept trying to figure out a way with Flexmock to accomplish the task.

https://github.com/borgmatic-collective/borgmatic/pull/86/commits/dab0dfcb32a6658f0d57309b0a87edc62bd825da